src/actions/requests.rs

```diff
@@ -217,69 +207,76 @@ impl<'a> Action<'a> for Definition {
     }
 }
 
+#[derive(Debug)]
+enum ResponseFrom<X, Y> {
+    Compiler(X),
+    Racer(Y),
+}
+
 impl<'a> RequestAction<'a> for Definition {
     type Response = Vec<Location>;
     fn handle<O: Output>(&mut self, _id: usize, params: Self::Params, ctx: &mut ActionContext, _out: O) -> Result<Self::Response, ()> {
         // Save-analysis thread.
-        let t = thread::current();
         let ctx = ctx.inited();
         let file_path = parse_file_path!(&params.text_document.uri, "goto_def")?;
         let span = ctx.convert_pos_to_span(file_path.clone(), params.position);
         let analysis = ctx.analysis.clone();
         let vfs = ctx.vfs.clone();
 
-        let compiler_handle = thread::spawn(move || {
-            let result = analysis.goto_def(&span);
-
-            t.unpark();
-
-            result
+        let (sender, receiver) = mpsc::channel();
```

Comment:
If we were to still [prioritize analysis data](https://github.com/rust-lang-nursery/rls/pull/552/files#r149239110), I think we could simplify this a great deal with `receive_from_thread`, like used elsewhere. We could start both threads, wait with timeout on the analysis one (this should succeed or fail fast in practice), and if needed wait for the remainder for the Racer one (iirc it can also take longer and is less reliable). It'd also remove the need to use `WORK_POOL` directly and do the explicit gymnastics with counters.

Reply:
Tbf I would have generalised the racing logic it if it were used more than once. 

As I wrote above I think you need to race the computations, or just use one or the other. If the compiler fails fast then it'll work fine in this case too. Except here we can handle a compiler-timeout and a fast _racer_ with grace.

I see that this is a change in functionality though, racer could never beat the compiler before. Let me know what you guys want to do with this.